### ISSUE TITLE
AttributeError when Appending "recall" History Data

### ISSUE THREAD

**Issue by ab-tools** [dscripka/openWakeWord#38]
Hello @dscripka,

one thing that happens sometimes and I'm not sure if that is actually an issue or I can just ignore it.

From time to time I get an `AttributeError` in this line:
`history['recall'].append(float(tp/(tp+fn).detach().numpy()))`

And the reason is that both `tp` and `fn` are `0` in that case.
As these variables are calculated from `predictions` I dumped that variable when this happens which looks like this as an example:
```
tensor([[2.6955e-06],
        [1.0686e-07],
        [5.5851e-05],
        [2.8651e-07],
        [4.0770e-07],
        [1.5586e-07],
        [1.5170e-07],
        [6.1524e-07],
        [2.2747e-07],
        [6.8342e-08],
        [2.7067e-07],
        [7.6920e-08],
        [2.1011e-07],
        [7.1434e-06],
        [1.5425e-06],
        [8.7802e-08],
        [1.1738e-07],
        [9.4471e-08],
        [1.2498e-07],
        [1.7970e-07],
        [2.1878e-07],
        [1.1008e-07],
        [1.5025e-06],
        [1.9297e-06],
        [3.6787e-07],
        [2.6216e-07],
        [3.7311e-07],
        [7.5499e-08],
        [1.0872e-07],
        [2.8575e-07],
        [1.3877e-07],
        [4.6585e-07],
        [9.3889e-08],
        [1.3185e-06],
        [1.8648e-06],
        [1.3281e-07],
        [1.8029e-07],
        [8.0312e-08],
        [1.1102e-07],
        [1.3540e-07],
        [2.5936e-07],
        [3.1800e-07],
        [6.2738e-06],
        [1.2455e-07],
        [1.9887e-07],
        [1.2764e-07],
        [4.9647e-07],
        [9.5518e-08],
        [6.5103e-07],
        [3.3730e-07],
        [1.1653e-07],
        [2.0794e-07],
        [1.5047e-07],
        [1.2003e-06],
        [1.9570e-07],
        [2.7696e-07],
        [2.2603e-07],
        [1.5527e-07],
        [1.6003e-07],
        [7.4593e-07],
        [1.3091e-07],
        [1.3737e-07],
        [1.2177e-07],
        [8.5175e-06],
        [2.0019e-07],
        [1.0543e-07],
        [2.2684e-07],
        [1.4450e-07],
        [1.0111e-07],
        [1.4446e-07],
        [8.3700e-08],
        [1.5691e-07],
        [9.6111e-08],
        [8.9532e-07],
        [6.0977e-07],
        [1.6099e-07],
        [7.6571e-07],
        [2.6326e-07],
        [1.1610e-07],
        [1.2645e-07],
        [1.0633e-07]], grad_fn=<SigmoidBackward0>)
```

Right now I just catch this exception and add a `0` to the `recall` history data when this happens:
```
        try:
            history['recall'].append(float(tp/(tp+fn).detach().numpy()))
        except AttributeError:
            print(predictions)
            history['recall'].append(float(0))
```

Did you see this issue already, too?
And is it safe to just ignore it like this or is there something wrong with the model?

Thanks again for all your support
Andreas

**Comment by dscripka** [comment]
Thank you for pointing this out, and I have seen this behavior intermittently in the past before. Ultimately it is due to the training data and model initialization, and generally only occurs near the beginning of the training before the model starts to learn. Your approach of inserting zeros is appropriate, and otherwise it is safe to ignore.

But you are correct that this situation should be caught by code, and it is something that will be fixed with the new training code and tutorial notebook.

**Comment by dscripka** [comment]
These are great questions @ab-tools, and training the models is definitely the most complex part of the process and part of the reason why it's taking me so long to create a training process that is more reproducible and should (hopefully) work across different types of wake words/phrases.

The repeatability issues across training runs is something I've noticed as well, and in practice there appear to be many different local minima that correspond to very different false positive rates. I have some approaches that I've found helpful in the past, but given that this is a very complex topic I've created a [discussion page](https://github.com/dscripka/openWakeWord/discussions/40) where we can talk about this in more detail. Setting `shuffle = True` in the data loader could be contributing to this as well (and it should be the only random factor, outside of model initialization and CUDA non-determinism), but I haven't tested that specifically.

Thank you for raising so many important questions about training new models; this is a very important feature for the library and hopefully with some more testing and community validation we can significantly improve the process.
